olive/passes/onnx/extract_adapters.py

```diff
@@ -0,0 +1,336 @@
+import logging
+import re
+from copy import deepcopy
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, List
+
+import numpy as np
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import model_proto_to_olive_model
+from olive.passes.onnx.onnx_dag import OnnxDAG
+from olive.passes.pass_config import PassConfigParam
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+
+logger = logging.getLogger(__name__)
+
+
+class ExtractAdapters(Pass):
+    """Extract adapter weights from model and save them as external weights file.
+
+    If make_inputs is False, model proto is invalid after this pass as the adapter weights point to non-existent
+    external files. Inference session must be created by first loading the adapter weights using
+    SessionOptions.add_external_initializers.
+
+    If make_inputs is True, the adapter weights are inputs to the model and must be provided during inference.
+    """
+
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "make_inputs": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description=(
+                    "Convert adapter weights to inputs. If false, the adapter weights will be set as initializers with"
+                    " external data."
+                ),
+            ),
+            "pack_inputs": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description=(
+                    "Pack adapter weights for the same module type into a single input tensor. Only used if make_inputs"
+                    " is True."
+                ),
+            ),
+        }
+
+    def _run_for_config(
+        self, model: ONNXModelHandler, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModelHandler:
+        output_model_path = resolve_onnx_path(output_model_path, Path(model.model_path).name)
+
+        if "lora_modules" not in model.model_attributes:
+            raise ValueError("Model does not contain lora_modules attribute")
+
+        # create a dag from the model
+        dag = OnnxDAG.from_model_path(model.model_path)
+        # remove unnecessary identity nodes
+        self.remove_identity_nodes(dag)
+
+        # get lora modules
+        lora_modules = model.model_attributes["lora_modules"]
+        lora_name_patterns = self._get_lora_name_patterns(lora_modules)
+
+        # dictionary to store adapter weights
+        weights = {}
+        # keep track of float and quantized modules
+        float_modules = set()
+        quant_modules = set()
+
+        # nodes to remove at the end
+        nodes_to_remove = set()
+
+        for node_name in list(dag.nodes.keys()):
```

Comment:
seems the loop only related to DAG and could be moved to the DAG class?

Reply:
the DAG object is meant to be a utils to represent an onnx model as a dag. It will only handle simple logic like adding, removing or replacing nodes. 
For example, the autofusion pass will also use the dag https://github.com/microsoft/Olive/blob/8ff3da1ff0ddc0b276156527712ad7801d3546a1/olive/passes/onnx/auto_fusion.py#L75

The loop here is specific to this pass. It will be responsible for the logic.